Lib/test/test_capi/test_tuple.py

```diff
@@ -0,0 +1,242 @@
+        self.assertEqual(getslice(tup, 0, 0), ())
+        self.assertEqual(getslice(tup, 3, 0), ())
+
+        # slice
+        tup = (1, [2], 3)
+        self.assertEqual(getslice(tup, 1, 3), ([2], 3))
+        tup = TupleSubclass((1, [2], 3))
+        self.assertEqual(getslice(tup, 1, 3), ([2], 3))
+
+        # whole
+        tup = (1, [2], 3)
+        self.assertEqual(getslice(tup, 0, 3), tup)
+        self.assertEqual(getslice(tup, 0, 100), tup)
+        self.assertEqual(getslice(tup, -100, 100), tup)
+        tup = TupleSubclass((1, [2], 3))
+        self.assertEqual(getslice(tup, 0, 3), tup)
+        self.assertEqual(getslice(tup, 0, 100), tup)
+        self.assertEqual(getslice(tup, -100, 100), tup)
+
+        self.assertRaises(SystemError, getslice, [1, [2], 3], 0, 1)
+        self.assertRaises(SystemError, getslice, 42, 0, 1)
+
+        # CRASHES getslice(NULL, 0, 0)
+
+    def test_tuple_setitem(self):
+        # Test PyTuple_SetItem()
+        setitem = _testlimitedcapi.tuple_setitem
+        checknull = _testcapi._check_item_is_NULL
+
+        tup = (0, [0])
+        self.assertEqual(setitem(tup, 0, [1]), ([1], [0]))
+        self.assertEqual(setitem(tup, 1, [1]), (0, [1]))
+        tup = setitem(tup, 1, NULL)
+        self.assertTrue(checknull(tup, 1))
+
+        self.assertRaises(IndexError, setitem, tup, PY_SSIZE_T_MIN, 1)
+        self.assertRaises(IndexError, setitem, tup, -1, 1)
+        self.assertRaises(IndexError, setitem, tup, len(tup), 1)
+        self.assertRaises(IndexError, setitem, tup, PY_SSIZE_T_MAX, 1)
+        self.assertRaises(SystemError, setitem, [1], 0, 1)
+        self.assertRaises(SystemError, setitem, 42, 0, 1)
+
+        # CRASHES setitem(NULL, 1, 5)
+
+    def test_tuple_set_item(self):
+        # Test PyTuple_SET_ITEM()
+        set_item = _testcapi.tuple_set_item
+        checknull = _testcapi._check_item_is_NULL
+
+        tup = (0, [0])
+        self.assertEqual(set_item(tup, 0, [1]), ([1], [0]))
+        self.assertEqual(set_item(tup, 1, [1]), (0, [1]))
+        tup = set_item(tup, 1, NULL)
+        self.assertTrue(checknull(tup, 1))
+
+    def test__tuple_resize(self):
+        # Test _PyTuple_Resize()
+        resize = _testcapi._tuple_resize
+        checknull = _testcapi._check_item_is_NULL
+
+        a = ()
+        b = resize(a, 0)
+        self.assertEqual(b, a)
+        b = resize(a, 2)
+        self.assertEqual(len(b), 2)
+
+        a = (1, [2], [3])
+        b = resize(a, 3)
+        self.assertEqual(b, a)
+        b = resize(a, 2)
+        self.assertEqual(b, a[:2])
+        b = resize(a, 5)
+        self.assertEqual(len(b), 5)
+        self.assertEqual(b[:3], a)
+        self.assertTrue(all(checknull(b, i) for i in range(3, 5)))
+
+        a = ()
+        self.assertRaises(MemoryError, resize, a, PY_SSIZE_T_MAX)
+        self.assertRaises(SystemError, resize, a, -1)
+        self.assertRaises(SystemError, resize, a, PY_SSIZE_T_MIN)
+        # refcount > 1
```

Comment:
i don't get how you make sure that the refcount is > 1 ? Maybe add self.assertGreater(sys.getrefcount(a), 1)?

Reply:
One reference from the local variable and other from a function argument.

Actually, it is > 1 in all other tests too.

Reply:
Ah ok, it makes sense.